src/v/kafka/server/handlers/describe_transactions.cc

```diff
@@ -0,0 +1,101 @@
+
+#include "cluster/errc.h"
+#include "cluster/tx_gateway_frontend.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/describe_transactions_request.h"
+#include "kafka/protocol/types.h"
+#include "kafka/server/handlers/details/security.h"
+#include "kafka/server/request_context.h"
+#include "kafka/server/response.h"
+#include "model/fundamental.h"
+#include "model/namespace.h"
+#include "resource_mgmt/io_priority.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/when_all.hh>
+
+namespace kafka {
+
+template<>
+ss::future<response_ptr> describe_transactions_handler::handle(
+  request_context ctx, ss::smp_service_group) {
+    describe_transactions_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+    vlog(klog.trace, "Handling request {}", request);
+
+    describe_transactions_response response;
+
+    auto unauthorized_it = std::partition(
+      request.data.transactional_ids.begin(),
+      request.data.transactional_ids.end(),
+      [&ctx](const kafka::transactional_id& tx_id) {
+          return ctx.authorized(security::acl_operation::describe, tx_id);
+      });
+
+    std::vector<kafka::transactional_id> unauthorized(
+      std::make_move_iterator(unauthorized_it),
+      std::make_move_iterator(request.data.transactional_ids.end()));
+
+    request.data.transactional_ids.erase(
+      unauthorized_it, request.data.transactional_ids.end());
+
+    auto& tx_frontend = ctx.tx_gateway_frontend();
+    if (likely(!request.data.transactional_ids.empty())) {
+        for (const auto& tx_id : request.data.transactional_ids) {
+            describe_transaction_state tx_info_resp;
+            tx_info_resp.transactional_id = tx_id;
+
+            auto tx_info = co_await tx_frontend.get_tx(tx_id);
+            if (tx_info.has_value()) {
+                tx_info_resp.producer_id = kafka::producer_id(
+                  tx_info.value().pid.id);
+                tx_info_resp.producer_epoch = tx_info.value().pid.get_epoch();
+                tx_info_resp.transaction_state = ss::sstring(
+                  tx_info.value().get_status());
+                tx_info_resp.transaction_timeout_ms
+                  = tx_info.value().get_timeout().count();
+                // Is it legal?
+                tx_info_resp.transaction_start_time_ms
+                  = tx_info.value().last_update_ts.time_since_epoch().count();
+            } else {
+                switch (tx_info.error()) {
+                case cluster::tx_errc::not_coordinator:
+                    tx_info_resp.error_code
+                      = kafka::error_code::not_coordinator;
+                    break;
+                case cluster::tx_errc::tx_id_not_found:
+                    tx_info_resp.error_code
+                      = kafka::error_code::transactional_id_not_found;
+                default:
+                    tx_info_resp.error_code
+                      = kafka::error_code::unknown_server_error;
```

Comment:
we should not return this.. per KIP following are acceptable.

```
NOT_COORDINATOR: If the coordinator receiving the request does not own a transactionalId in the request.
COORDINATOR_LOAD_IN_PROGRESS: The coordinator is in the process of loading its state.
COORDINATOR_NOT_AVAILABLE: If the coordinator receiving the request is being shutdown.
TRANSACTIONAL_ID_NOT_FOUND (NEW): New error code which indicates that the requested TransactionalId could not be found 
TRANSACTIONAL_ID_AUTHORIZATION_FAILED: If the principal does not have Describe permission one of the transactionalIds in the request.
```

Reply:
We can return it
We need return smth fatal for client when smth goes wrong. Because another type of error client can handle and do some retries, for example

Reply:
As I mentioned in my other comment, we have been bitten once by returning a a default unknown_server_once. if you can double check the client code and make sure it handles USE gracefully, think it is fine, other wise it should be a last resort IMO.